File: server/src/main/java/com/fluffy/submission/domain/Submission.java
package com.fluffy.submission.domain;

import com.fluffy.global.persistence.AuditableEntity;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import java.util.ArrayList;
import java.util.List;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Getter
public class Submission extends AuditableEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private Long examId;

    @Column(nullable = false)
    private Long memberId;

    @OneToMany(
            mappedBy = "submission",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    private final List<Answer> answers = new ArrayList<>();

    public Submission(Long examId, Long memberId, List<Answer> answers) {
        this(null, examId, memberId, new ArrayList<>(answers));
    }

    public Submission(Long id, Long examId, Long memberId, List<Answer> answers) {
        this.id = id;
        this.examId = examId;
        this.memberId = memberId;
        addAnswers(answers);
    }

    private void addAnswers(List<Answer> answers) {
        answers.forEach(answer -> answer.updateSubmission(this));
        this.answers.addAll(answers);
    }

    public boolean isNotWrittenBy(Long memberId) {
        return !this.memberId.equals(memberId);
    }
}

File: server/src/main/kotlin/com/fluffy/submission/domain/Submission.kt
package com.fluffy.submission.domain

import com.fluffy.infra.persistence.AuditableEntity
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.OneToMany

@Entity
class Submission private constructor(
    examId: Long,
    memberId: Long,
    id: Long = 0,
) : AuditableEntity() {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = id
        protected set

    @field:Column(nullable = false)
    var examId: Long = examId
        protected set

    @field:Column(nullable = false)
    var memberId: Long = memberId
        protected set

    @OneToMany(mappedBy = "submission", cascade = [CascadeType.ALL], orphanRemoval = true)
    private val _answers: MutableList<Answer> = mutableListOf()
    val answers: List<Answer> get() = _answers.toList()

    companion object {
        fun create(examId: Long, memberId: Long, answers: List<Answer>): Submission {
            val submission = Submission(examId, memberId)
            submission.addAnswers(answers)

            return submission
        }
    }

    fun addAnswers(answers: List<Answer>) {
        answers.forEach { it.updateSubmission(this) }
        _answers.addAll(answers)
    }

    fun isNotWrittenBy(memberId: Long): Boolean {
        return this.memberId != memberId
    }
}